src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+            sol_mqtt_event_loop, mqtt);
+
+    return 0;
+}
+
+SOL_API void
+sol_mqtt_disconnect(struct sol_mqtt *mqtt)
+{
+    CHECK_INIT();
+    SOL_NULL_CHECK(mqtt);
+
+    if (mqtt->socket_read)
+        sol_fd_del(mqtt->socket_read);
+    if (mqtt->socket_write)
+        sol_fd_del(mqtt->socket_write);
+
+    if (mqtt->connect_timeout)
+        sol_timeout_del(mqtt->connect_timeout);
+    if (mqtt->disconnect_timeout)
+        sol_timeout_del(mqtt->disconnect_timeout);
+    if (mqtt->publish_timeout)
+        sol_timeout_del(mqtt->publish_timeout);
+    if (mqtt->message_timeout)
+        sol_timeout_del(mqtt->message_timeout);
+    if (mqtt->subscribe_timeout)
+        sol_timeout_del(mqtt->subscribe_timeout);
+    if (mqtt->unsubscribe_timeout)
+        sol_timeout_del(mqtt->unsubscribe_timeout);
+
+    mosquitto_disconnect_callback_set(mqtt->mosq, NULL);
+    mosquitto_disconnect(mqtt->mosq);
+
+    mosquitto_destroy(mqtt->mosq);
+    free(mqtt);
+
+    sol_mqtt_shutdown();
+}
+
+SOL_API int
+sol_mqtt_get_connection_status(struct sol_mqtt *mqtt)
+{
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    return mqtt->connection_status;
+}
+
+SOL_API int
+sol_mqtt_publish(struct sol_mqtt *mqtt, struct sol_mqtt_message *message)
+{
+    int r;
+
+    CHECK_INIT(-EINVAL);
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    r = mosquitto_publish(mqtt->mosq, &message->id, message->topic, message->payload_len, message->payload,
+            (int) message->qos, message->retain);
+
+    if (r != MOSQ_ERR_SUCCESS) {
+        SOL_WRN("Unable to publish to '%s'", message->topic);
+        return -EINVAL;
+    }
+
+    if (!mqtt->socket_write) {
+        mqtt->socket_write = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_OUT, sol_mqtt_event_loop, mqtt);
+        SOL_NULL_CHECK(mqtt->socket_write, -EINVAL);
+    }
+
+    return 0;
+}
+
+SOL_API int
+sol_mqtt_subscribe(struct sol_mqtt *mqtt, char *topic, size_t topic_len, sol_mqtt_qos qos)
+{
+    int r;
+    char *s_topic;
+
+    CHECK_INIT(-EINVAL);
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    SOL_NULL_CHECK(topic, -EINVAL);
+
+    s_topic = strndupa(topic, topic_len);
+
+    r = mosquitto_subscribe(mqtt->mosq, NULL, s_topic, (int) qos);
+    if (r != MOSQ_ERR_SUCCESS) {
+        SOL_WRN("Unable to subscribe to '%s'", topic);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+SOL_API struct sol_mqtt_message *
+sol_mqtt_message_new(char *topic, size_t topic_len, void *payload, size_t payload_len,
+        sol_mqtt_qos qos, bool retain)
+{
+    struct sol_mqtt_message *message;
+
+    SOL_NULL_CHECK(topic, NULL);
+    SOL_NULL_CHECK(payload, NULL);
+
+    message = calloc(1, sizeof(struct sol_mqtt_message));
+    SOL_NULL_CHECK(message, NULL);
+
+    message->topic = calloc(1, topic_len + 1);
+    SOL_NULL_CHECK_GOTO(message->topic, topic_error);
+    memcpy(message->topic, topic, topic_len);
+
+    message->payload = sol_util_memdup(payload, payload_len);
```

Comment:
this is bad. What if we use `sol_blob` and take a reference? Then we release the reference whe it's not needed anymore.